core/src/ledger/storage/mod.rs

```diff
@@ -1182,6 +1193,52 @@ where
             .unwrap_or_default()
     }
 
+    /// Get the oldest epoch where we can read a value
+    pub fn get_oldest_epoch(&self) -> Epoch {
+        let oldest_height = match self.storage_read_past_height_limit {
+            Some(limit) if limit < self.get_last_block_height().0 => {
+                (self.get_last_block_height().0 - limit).into()
+            }
+            _ => BlockHeight(1),
+        };
+        self.block
+            .pred_epochs
+            .get_epoch(oldest_height)
+            .unwrap_or_default()
+    }
+
+    /// Get oldest epoch which has the valid signed nonce of the bridge pool
+    pub fn get_oldest_epoch_with_valid_nonce(&self) -> Result<Epoch> {
+        let nonce_key = bridge_pool::get_nonce_key();
```

Comment:
this is incorrect. `get_nonce_key` doesn't return a signed nonce; rather, it returns the latest Bridge pool nonce, which may lag behind. the key you want is `get_signed_bridge_pool_root`, which returns a tuple containing a `BridgePoolRootProof` as one of the elements. inside of `BridgePoolRootProof` there will be a `Uint` corresponding to the signed nonce.

Reply:
Can we move `BridgePoolRootProof` and other types from `ethereum_bridge` to `core` to decode `BridgePoolRootProof` in `core` storage?